node.go

```diff
@@ -135,20 +140,95 @@ func (n *Node) EnableMDNSDiscovery(notifee discovery.Notifee, interval time.Dura
 	return
 }
 
+func (n *Node) ExternalAddr() ma.Multiaddr {
+	if n.nat == nil {
+		return n.NetAddr
+	} else {
+		mappings := n.nat.Mappings()
+		for i := 0; i < len(mappings); i++ {
+			external_addr, err := mappings[i].ExternalAddr()
+			if err == nil {
+				return external_addr
+			}
+		}
+		return n.NetAddr
+	}
+}
+
+func (n *Node) discoverAndHandleNat(listenPort int) {
+	Debugf("Looking for a NAT...")
+	n.nat = nat.DiscoverNAT()
+	if n.nat == nil {
+		Debugf("No NAT found.")
+	} else {
+		Debugf("Discovered NAT! Trying to aquire public port mapping via UPnP...")
+		ifaces, _ := go_net.Interfaces()
+		// handle err
+		for _, i := range ifaces {
+			addrs, _ := i.Addrs()
+			// handle err
+			for _, addr := range addrs {
+				var ip go_net.IP
+				switch v := addr.(type) {
+				case *go_net.IPNet:
+					ip = v.IP
+				case *go_net.IPAddr:
+					ip = v.IP
+				}
+				if ip.Equal(go_net.IPv4(127, 0, 0, 1)) {
+					continue
+				}
+				addr_string := fmt.Sprintf("/ip4/%s/tcp/%d", ip, listenPort)
+				localaddr, err := ma.NewMultiaddr(addr_string)
+				if err == nil {
+					Debugf("NAT: trying to establish NAT mapping for %s...", addr_string)
+					n.nat.NewMapping(localaddr)
+				}
+			}
+		}
+
+		external_addr := n.ExternalAddr()
+
+		if external_addr != n.NetAddr {
+			Debugf("NAT: successfully created port mapping! External address is: %s", external_addr.String())
+		} else {
+			Debugf("NAT: could not create port mappping. Keep trying...")
+			fmt.Errorf("NAT:-------------------------------------------------------")
```

Comment:
maybe this should use the Info log, not just standard err?  I have a function Infof().  Maybe we should have a longer conversation about the logs and their usage, but I don't think I use Errorf anywhere else and we should unify things.

Reply:
Ok, will change to Infof().